File: 0234-palindrome-linked-list/0234-palindrome-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def isPalindrome(self, head: Optional[ListNode]) -> bool:
        #### using list[::-1]  
        if not head:
            return True
        val_list = []
        while head:
            val_list.append(head.val)
            head = head.next
        return val_list == val_list[::-1]


        #### using deque
        #if not head:
        #    return True
        #from collections import deque
        #q = deque([head.val])
        #while head.next:
        #    q.append(head.next.val)
        #    head = head.next
        #flag = True
        #while flag and len(q) >= 2:
        #    if q.popleft() != q.pop():
        #        return False
        #return True

File: 0234-palindrome-linked-list/0234-palindrome-linked-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    bool isPalindrome(ListNode* head) {
        ListNode* slow = head;
        ListNode* fast = head;
        while (fast != nullptr && fast->next != nullptr) {
            slow = slow->next;
            fast = fast->next->next;
        }
        ListNode* rev = reverse(slow); 
        while (rev != nullptr) {
            if (head->val != rev->val) return false;
            head = head->next;
            rev = rev->next;
        }
        return true;
    }
private:
    ListNode* reverse(ListNode* node) {
        ListNode* prev = nullptr;
        ListNode* curr = node;
        while (curr != nullptr) {
            ListNode* next = curr->next;
            prev = curr;
            curr = next;
        }
        return prev;
    }
};